src/main/rascal/examples/pico/Rename.rsc

```diff
@@ -41,93 +41,46 @@ import util::FileSystem;
 
 data Tree;
 
-alias RenameRequest = tuple[list[Tree] cursor, str newName/*, set[loc] workspaceFolders*/];
-
-data RenameState
-    = findDefinition(RenameRequest req)
-    | rename(IdRole role, loc def, str newName)
-    ;
-
-@memo{expireAfter(minutes=1), maximumSize(50)}
-Tree parseLoc(loc l) {
-    return parse(#start[Program], l);
-}
-
-@memo{expireAfter(minutes=1), maximumSize(50)}
-TModel tmodelForLoc(loc l) {
-    return collectAndSolve(parseLoc(l));
-}
+public tuple[list[DocumentEdit] edits, map[str, ChangeAnnotation] annos, set[Message] msgs] renamePico(list[Tree] cursor, str newName) {
+    if (!isValidName(newName)) {
+        return <[], (), {error("\'<newName>\' is not a valid name here.", cursor[0].src)}>;
+    }
 
-public tuple[list[DocumentEdit] edits, map[str, ChangeAnnotation] annos, set[Message] msgs] renamePico(RenameRequest request) {
-    RenameConfig config = rconfig(
-        parseLoc
-      , tmodelForLoc
-      , reportCollectCycles = true
+    return rename(
+        cursor
+      , newName
+      , Tree(loc l) { return parse(#start[Program], l); }
+      , collectAndSolve
+      , findDefinitions
+      , findCandidateFiles
+      , renameDef
     );
-
-    RenameSolver solver = newSolverForConfig(config);
-    initSolver(solver, request);
-    return solver.run();
 }
 
-void initSolver(RenameSolver solver, RenameRequest req) {
-    if (!isValidName(req.newName)) {
-        // gooi errors
-        solver.msg(error("Not a valid name: \'<req.newName>\'!"));
-        return;
+set[Define] findDefinitions(list[Tree] cursor, TModel(Tree) getTModel, Renamer r) {
+    TModel tm = getTModel(cursor[-1]);
+    if (Tree t <- cursor
+      , tm.definitions[t.src]?) {
+            return {tm.definitions[t.src]};
     }
 
-    loc currentModule = req.cursor[0].src.top;
-    solver.collectTModel(currentModule, renameByModel, findDefinition(req));
-}
-
-void renameByModel(findDefinition(RenameRequest req), TModel tm, RenameSolver solver) {
-    println("Finding definition in TModel \'<tm.modelName>\'");
-
-    loc fileUnderCursor = req.cursor[0].src.top;
-    loc def = getDefLocation(tm, req.cursor);
-    IdRole defRole = tm.definitions[def].idRole;
-
-    // Rename occurrences in this file
-    solver.collectTModel(fileUnderCursor, renameByModel, rename(defRole, def, req.newName));
-
-    // Rename occurrences in any other files, if the renaming is not local
-    // if (!isLocalRename(tm, def)) {
-    //     for (loc m <- possibleDownstreamUsers(solver, req.workspaceFolders, def), m != fileUnderCursor) {
-    //         solver.collectParseTree(m, renameByTree, rename(defRole, def, req.newName));
-    //     }
-    // }
+    r.msg(error("No definition for name under cursor", cursor[0].src));
```

Comment:
we could also add overloads for `error`/`warning`/`info` next to this `msg`, just to help the user out.

possible include the format support we have in typepal.

Reply:
and shouldn't this be an error message that the frameworkr generates automatically if you return an empty set (or throw an exception in this functiion)